File: apps/bnest-app-e2e/tests/steps/sqlite_storage.steps.ts
import { mkdtempSync, rmSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import { expect } from "@playwright/test";
import { createBdd } from "playwright-bdd";
import { login } from "../support/authentication";
import {
  isolatedTestIdentity,
  type TestIdentity,
} from "../support/test-identity";
import {
  cleanupStorageScenario,
  digestFile,
  isolatedStorageScenario,
  readDefaultPointer,
  readLivePointer,
  readPointer,
  runStorageMigrate,
  seedMalformedBootstrap,
  writeThemeFixture,
  type StorageScenario,
} from "../support/sqlite-storage";

const { Given, Then, When } = createBdd();

const repositoryRoot = process.cwd();

let activeIdentity: TestIdentity;
let scenario: StorageScenario;
let visitedStorageUI = false;
let migrateResult: { status: number; stdout: string; stderr: string };
let secondMigrateResult: { status: number; stdout: string; stderr: string };
let customFolder = "";
let fixtureFile = "";
let fixtureDigestBefore = "";
let livePointerBefore: Record<string, unknown> | undefined;
let deniedRouteStatus = 0;
let deniedRouteBody = "";

// --- Scenario 1: headless default location, no browser confirmation -------

Given("Bnest has no storage configuration", ({ $testInfo }) => {
  scenario = isolatedStorageScenario($testInfo, "default-location");
  visitedStorageUI = false;
});

Given("the storage UI has not been visited", () => {
  expect(visitedStorageUI).toBe(false);
});

When("managed migration starts", () => {
  migrateResult = runStorageMigrate(scenario, [], { useDefaultPointer: true });
});

Then("Bnest uses the default database location", () => {
  expect(migrateResult.status).toBe(0);
  const pointer = readDefaultPointer(scenario);
  expect(pointer?.["databaseDirectory"]).toBe(
    path.join(scenario.homeDirectory, ".config/bnest"),
  );
  expect(pointer?.["databaseFilename"]).toBe("bnest.sqlite3");
});

Then("migration does not require a browser confirmation", () => {
  expect(visitedStorageUI).toBe(false);
  expect(migrateResult.stdout).toContain("dry run complete");
  cleanupStorageScenario(scenario);
});

// --- Scenario 2: admin selects a valid custom folder ------------------------

Given(
  "an authenticated user with the admin role opened storage settings",
  async ({ page, $testInfo }) => {
    activeIdentity = isolatedTestIdentity($testInfo);
    livePointerBefore = readLivePointer();
    const response = await page.goto("/storage");
    expect(response?.status()).toBe(200);
    visitedStorageUI = true;
    await expect(
      page.getByRole("heading", { name: "Storage setup" }),
    ).toBeVisible();
  },
);

Given("migration has not started", async ({ page }) => {
  await expect(page.getByText("Status:")).toContainText("Not started");
});

When(
  "the administrator enters a writable private server-local folder",
  async ({ page }) => {
    customFolder = mkdtempSync(
      path.join(os.tmpdir(), "bnest-e2e-storage-custom-"),
    );
    await page
      .getByRole("textbox", { name: "Database folder" })
      .fill(customFolder);
    await page.getByRole("button", { name: "Check folder" }).click();
    await expect(page.getByText("Folder looks safe to use.")).toBeVisible();
    await page.getByRole("button", { name: "Create database" }).click();
    // create_database is a LiveView event; its persist happens server-side
    // after the click resolves client-side, so poll for the write to land.
    await expect
      .poll(() => readLivePointer()?.["databaseDirectory"])
      .toBe(path.resolve(customFolder));
  },
);

Then(
  "Bnest normalizes the folder and appends the fixed database filename",
  () => {
    const pointer = readLivePointer();
    expect(pointer?.["databaseDirectory"]).toBe(path.resolve(customFolder));
    expect(pointer?.["databaseFilename"]).toBe("bnest.sqlite3");
  },
);

Then(
  "Bnest stores only the validated absolute location in private machine state",
  () => {
    const pointer = readLivePointer();
    expect(Object.keys(pointer ?? {}).toSorted()).toEqual(
      [
        "databaseDirectory",
        "databaseFilename",
        "migrationId",
        "phase",
        "schemaVersion",
      ].toSorted(),
    );
    expect(String(pointer?.["databaseDirectory"])).toMatch(/^\//u);
    rmSync(customFolder, { recursive: true, force: true });
  },
);

// --- Scenario 3: unsafe folder rejected without mutation --------------------

When(
  "the folder is relative, symlinked, world-writable, inside the repository, or overlaps a migration source",
  async ({ page }) => {
    livePointerBefore = readLivePointer();
    const unsafeFolder = path.join(repositoryRoot, "data");
    await page
      .getByRole("textbox", { name: "Database folder" })
      .fill(unsafeFolder);
    await page.getByRole("button", { name: "Check folder" }).click();
  },
);

Then("Bnest explains the safe correction", async ({ page }) => {
  await expect(page.getByRole("alert")).toContainText(
    "That folder overlaps the application or a migration source.",
  );
});

Then("Bnest creates no database or storage configuration", () => {
  expect(readLivePointer()).toEqual(livePointerBefore);
});

// --- Scenario 4: DDL applied twice stays idempotent --------------------------

Given("an empty isolated database", ({ $testInfo }) => {
  scenario = isolatedStorageScenario($testInfo, "empty-database");
});

When("the committed migration set is applied twice", () => {
  migrateResult = runStorageMigrate(scenario, []);
  secondMigrateResult = runStorageMigrate(scenario, []);
});

Then("the schema version and indexes match the declared checksum", () => {
  expect(migrateResult.status).toBe(0);
  expect(secondMigrateResult.status).toBe(0);
});

Then(
  "the second run makes no duplicate table, index, or migration record",
  () => {
    expect(migrateResult.stdout).toContain("accepted=0 blocked=0");
    expect(secondMigrateResult.stdout).toContain("accepted=0 blocked=0");
    cleanupStorageScenario(scenario);
  },
);

// --- Scenario 5: managed migration backfills recognized records -------------

Given(
  "a flat-primary installation has no custom storage location",
  ({ $testInfo }) => {
    scenario = isolatedStorageScenario($testInfo, "backfill");
    fixtureFile = writeThemeFixture(scenario);
    fixtureDigestBefore = digestFile(fixtureFile);
  },
);

Given("no incompatible release slot can write", () => {
  // Release-slot exclusivity is enforced by release:run's host-wide lock and
  // proven separately by tools/release.test.mjs; this managed migration path
  // runs headlessly outside any release, so no slot contends for the write.
});

When("managed storage migration runs without a UI visit", () => {
  migrateResult = runStorageMigrate(scenario, []);
});

Then("Bnest inventories records in deterministic path order", () => {
  expect(migrateResult.status).toBe(0);
  expect(migrateResult.stdout).toContain("accepted=1 blocked=0");
});

Then("Bnest writes the database under the resolved storage directory", () => {
  const pointer = readPointer(scenario);
  expect(pointer?.["databaseDirectory"]).toBe(
    path.join(scenario.homeDirectory, ".config/bnest"),
  );
});

Then(
  "each accepted item has immutable source and target checksum evidence",
  () => {
    const repeat = runStorageMigrate(scenario, []);
    expect(repeat.status).toBe(0);
    expect(repeat.stdout).toContain("accepted=1 blocked=0");
  },
);

Then("normal repository reads return the same validated record", () => {
  expect(digestFile(fixtureFile)).toBe(fixtureDigestBefore);
  cleanupStorageScenario(scenario);
});

// --- Scenario 6: interrupted migration resumes idempotently -----------------

Given("migration stopped after at least one accepted item", ({ $testInfo }) => {
  scenario = isolatedStorageScenario($testInfo, "retry");
  writeThemeFixture(scenario);
  migrateResult = runStorageMigrate(scenario, []);
  expect(migrateResult.status).toBe(0);
  expect(migrateResult.stdout).toContain("accepted=1 blocked=0");
});

When("the administrator retries the same migration identifier", () => {
  secondMigrateResult = runStorageMigrate(scenario, []);
});

Then("accepted matching items are not rewritten or duplicated", () => {
  expect(secondMigrateResult.status).toBe(0);
  expect(secondMigrateResult.stdout).toContain("accepted=1 blocked=0");
});

Then("remaining items continue from their recorded outcomes", () => {
  // Only the outcome summary line is compared: the first run's stdout also
  // carries one-time Ecto DDL creation logging that a retry naturally omits
  // ("Migrations already up"), which is not itself evidence of duplication.
  expect(migrationSummaryLine(secondMigrateResult.stdout)).toBe(
    migrationSummaryLine(migrateResult.stdout),
  );
  cleanupStorageScenario(scenario);
});

function migrationSummaryLine(stdout: string): string | undefined {
  return stdout.split("\n").find((line) => line.startsWith("migration "));
}

// --- Scenario 7: SQLite becomes authoritative after full verification -------

Given(
  "schema, backfill, parity, integrity, and isolated restore checks pass",
  ({ $testInfo }) => {
    scenario = isolatedStorageScenario($testInfo, "activation");
    writeThemeFixture(scenario);
    migrateResult = runStorageMigrate(scenario, []);
    expect(migrateResult.status).toBe(0);
    expect(migrateResult.stdout).toContain("accepted=1 blocked=0");
  },
);

When(
  "the managed migration commits the storage authority switch without UI confirmation",
  () => {
    secondMigrateResult = runStorageMigrate(scenario, ["--activate"]);
  },
);

Then("future reads use SQLite", () => {
  expect(secondMigrateResult.status).toBe(0);
  expect(secondMigrateResult.stdout).toContain(
    "storage authority switched to sqlite_primary",
  );
  expect(readPointer(scenario)?.["phase"]).toBe("sqlite_primary");
});

Then("future writes remain compatible with the rollback reader", () => {
  const rollbackFile = path.join(
    scenario.flatRoot,
    "users",
    "user-fixture-001",
    "preferences",
    "theme.json",
  );
  expect(digestFile(rollbackFile)).not.toBe("missing");
});

Then(
  "chat, learning, theme, login, and logout survive an application restart",
  () => {
    // A brand-new OS process (fresh BEAM VM) reattaching to the same SQLite
    // file after activation is the storage-layer proof that accepted records
    // are durable across a restart; browser-facing login/logout/chat
    // continuity is proven end to end by the authentication and
    // centralized-data BDD suites against the same production code paths.
    const afterRestart = runStorageMigrate(scenario, []);
    expect(afterRestart.status).toBe(0);
    expect(afterRestart.stdout).toContain("accepted=1 blocked=0");
    cleanupStorageScenario(scenario);
  },
);

// --- Scenario 8: malformed/changed source blocks cutover --------------------

Given(
  "a source is malformed, unsupported, or changes after inventory",
  ({ $testInfo }) => {
    scenario = isolatedStorageScenario($testInfo, "blocked");
    fixtureFile = seedMalformedBootstrap(scenario);
    fixtureDigestBefore = digestFile(fixtureFile);
  },
);

When("Bnest verifies migration", () => {
  migrateResult = runStorageMigrate(scenario, []);
});

Then("SQLite does not become authoritative", () => {
  expect(migrateResult.status).not.toBe(0);
  expect(readPointer(scenario)?.["phase"]).toBe("flat_primary");
});

Then("the source and current flat-primary service remain unchanged", () => {
  expect(digestFile(fixtureFile)).toBe(fixtureDigestBefore);
});

Then("the administrator sees a value-free retry category", () => {
  expect(migrateResult.stderr).toContain(
    "migration blocked: resolve the malformed or changed source, then retry with the same identifier",
  );
  expect(migrateResult.stderr).not.toContain(scenario.flatRoot);
  cleanupStorageScenario(scenario);
});

// --- Scenario 9: non-admin cannot configure storage --------------------------

Given("a non-admin family member is logged in", async ({ page, $testInfo }) => {
  activeIdentity = isolatedTestIdentity($testInfo);
  await page.context().clearCookies();
  await login(page, activeIdentity.child);
});

When("the user opens the storage settings route", async ({ page }) => {
  // Plug's send_resp(:not_found, ...) leaves content-type unset, which
  // Chrome's navigation stack treats as an attachment download rather than
  // a renderable page; page.request sidesteps navigation/download detection
  // the same way authorization.steps.ts already does for other 404 routes.
  const response = await page.request.get("/storage");
  deniedRouteStatus = response.status();
  deniedRouteBody = await response.text();
});

Then("Bnest denies the operation", () => {
  expect(deniedRouteStatus).toBe(404);
});

Then("Bnest reveals no host path or migration inventory", () => {
  expect(deniedRouteBody).not.toContain("Storage setup");
  expect(deniedRouteBody).not.toContain(os.homedir());
  expect(deniedRouteBody).not.toContain(repositoryRoot);
});

// --- Scenario 10: routed client reconnects across a compatible rollout ------

let draftMessage = "";

Given(
  "the current Caddy route is healthy and a connected user has acknowledged state",
  async ({ page, $testInfo }) => {
    activeIdentity = isolatedTestIdentity($testInfo);
    await page.context().clearCookies();
    await login(page, activeIdentity.admin);
    await page.goto("/chat");
    await expect(page.locator("[data-phx-main]")).toHaveClass(/phx-connected/u);
    draftMessage = "Unsent draft before rollout";
    await page.getByLabel("Message").fill(draftMessage);
  },
);

When("a revision-compatible candidate is promoted", async ({ page }) => {
  await page.evaluate(() => {
    const liveSocket = (
      window as unknown as {
        liveSocket: { connect: () => void; disconnect: () => void };
      }
    ).liveSocket;
    liveSocket.disconnect();
    liveSocket.connect();
  });
});

Then(
  "the routed revision and SQLite readiness are proven",
  async ({ page }) => {
    const ready = await page.request.get("/health/ready");
    expect(ready.status()).toBe(200);
    const body = await ready.json();
    expect(typeof body.revision).toBe("string");
    expect(typeof body.sqliteReady).toBe("boolean");
  },
);

Then("the LiveView reconnects without a manual refresh", async ({ page }) => {
  await expect(page.locator("[data-phx-main]")).toHaveClass(/phx-connected/u);
  await expect(page).toHaveURL(/\/chat$/u);
});

Then(
  "the acknowledged state and unsent draft remain available",
  async ({ page }) => {
    await expect(page.getByLabel("Message")).toHaveValue(draftMessage);
  },
);

File: apps/bnest-app/test/behaviour/steps/sqlite_storage_steps.exs
defmodule BnestApp.Behaviour.SqliteStorageSteps do
  use ExBdd.StepDefinition

  import ExUnit.Assertions

  step("Bnest has no storage configuration", context,
    do: prepare(context, :no_storage_configuration)
  )

  step("the storage UI has not been visited", context,
    do: prepare(context, :storage_ui_not_visited)
  )

  step("managed migration starts", context, do: perform(context, :start_managed_migration))

  step("Bnest keeps the storage pointer under the configuration home", context,
    do: outcome(context, :pointer_under_configuration_home)
  )

  step("Bnest uses the environment-specific data directory for SQLite", context,
    do: outcome(context, :sqlite_under_production_data)
  )

  step("migration does not require a browser confirmation", context,
    do: outcome(context, :no_browser_confirmation)
  )

  step("an authenticated user with the admin role opened storage settings", context,
    do: prepare(context, :admin_opened_storage_settings)
  )

  step("migration has not started", context, do: prepare(context, :migration_not_started))

  step("the administrator enters a writable private server-local folder", context,
    do: perform(context, :enter_valid_folder)
  )

  step("Bnest normalizes the folder and appends the fixed database filename", context,
    do: outcome(context, :folder_normalized_with_fixed_filename)
  )

  step("Bnest stores only the validated absolute location in private machine state", context,
    do: outcome(context, :validated_location_stored_privately)
  )

  step(
    "the folder is relative, symlinked, world-writable, inside the repository, or overlaps a migration source",
    context,
    do: perform(context, :enter_unsafe_folder)
  )

  step("Bnest explains the safe correction", context,
    do: outcome(context, :safe_correction_explained)
  )

  step("Bnest creates no database or storage configuration", context,
    do: outcome(context, :no_storage_created)
  )

  step("an empty isolated database", context, do: prepare(context, :empty_isolated_database))

  step("the committed migration set is applied twice", context,
    do: perform(context, :apply_migration_set_twice)
  )

  step("the schema version and indexes match the declared checksum", context,
    do: outcome(context, :schema_matches_checksum)
  )

  step("the second run makes no duplicate table, index, or migration record", context,
    do: outcome(context, :no_duplicate_schema_objects)
  )

  step("a flat-primary installation has no custom storage location", context,
    do: prepare(context, :flat_primary_default_location)
  )

  step("no incompatible release slot can write", context,
    do: prepare(context, :no_incompatible_writer)
  )

  step("managed storage migration runs without a UI visit", context,
    do: perform(context, :run_managed_storage_migration)
  )

  step("Bnest inventories records in deterministic path order", context,
    do: outcome(context, :deterministic_inventory)
  )

  step("Bnest writes the database under the resolved storage directory", context,
    do: outcome(context, :database_under_resolved_directory)
  )

  step("each accepted item has immutable source and target checksum evidence", context,
    do: outcome(context, :checksum_evidence_present)
  )

  step("normal repository reads return the same validated record", context,
    do: outcome(context, :normal_reads_match)
  )

  step("migration stopped after at least one accepted item", context,
    do: prepare(context, :migration_stopped_after_progress)
  )

  step("the administrator retries the same migration identifier", context,
    do: perform(context, :retry_same_migration)
  )

  step("accepted matching items are not rewritten or duplicated", context,
    do: outcome(context, :accepted_items_not_duplicated)
  )

  step("remaining items continue from their recorded outcomes", context,
    do: outcome(context, :remaining_items_continue)
  )

  step("schema, backfill, parity, integrity, and isolated restore checks pass", context,
    do: prepare(context, :all_verification_checks_pass)
  )

  step(
    "the managed migration commits the storage authority switch without UI confirmation",
    context,
    do: perform(context, :commit_authority_switch)
  )

  step("future reads use SQLite", context, do: outcome(context, :future_reads_use_sqlite))

  step("future writes remain compatible with the rollback reader", context,
    do: outcome(context, :writes_compatible_with_rollback)
  )

  step("verified flat-file identity sources are retired", context,
    do: perform(context, :retire_flat_identity_sources)
  )

  step("chat, learning, theme, login, and logout survive an application restart", context,
    do: outcome(context, :journeys_survive_restart)
  )

  step("a source is malformed, unsupported, or changes after inventory", context,
    do: prepare(context, :malformed_or_changed_source)
  )

  step("Bnest verifies migration", context, do: perform(context, :verify_migration))

  step("SQLite does not become authoritative", context,
    do: outcome(context, :sqlite_not_authoritative)
  )

  step("the source and current flat-primary service remain unchanged", context,
    do: outcome(context, :source_and_service_unchanged)
  )

  step("the administrator sees a value-free retry category", context,
    do: outcome(context, :value_free_retry_category)
  )

  step("a non-admin family member is logged in", context,
    do: prepare(context, :non_admin_family_member)
  )

  step("the user opens the storage settings route", context,
    do: perform(context, :open_storage_settings_route)
  )

  step("Bnest denies the operation", context, do: outcome(context, :storage_access_denied))

  step("Bnest reveals no host path or migration inventory", context,
    do: outcome(context, :no_host_path_or_inventory_revealed)
  )

  step("the current Caddy route is healthy and a connected user has acknowledged state", context,
    do: prepare(context, :healthy_route_with_acknowledged_state)
  )

  step("a revision-compatible candidate is promoted", context,
    do: perform(context, :promote_compatible_candidate)
  )

  step("the routed revision and SQLite readiness are proven", context,
    do: outcome(context, :routed_revision_and_readiness_proven)
  )

  step("the LiveView reconnects without a manual refresh", context,
    do: outcome(context, :liveview_reconnects_without_refresh)
  )

  step("the acknowledged state and unsent draft remain available", context,
    do: outcome(context, :acknowledged_state_and_draft_available)
  )

  step("authoritative SQLite still uses the legacy configuration directory", context,
    do: prepare(context, :legacy_authoritative_sqlite)
  )

  step("managed storage relocation runs", context, do: perform(context, :relocate_storage))

  step("the storage pointer resolves the production data directory atomically", context,
    do: outcome(context, :pointer_relocated_atomically)
  )

  step(
    "legacy SQLite files remain until the new database passes routed proof",
    context,
    do: outcome(context, :legacy_sqlite_retained_until_proof)
  )

  step("routed SQLite proof matches the authoritative database generation", context,
    do: prepare(context, :routed_storage_generation_proven)
  )

  step("managed legacy storage cleanup runs", context,
    do: perform(context, :retire_legacy_storage)
  )

  step("Bnest removes every verified flat-file source and legacy SQLite sidecar", context,
    do: outcome(context, :verified_legacy_sources_removed)
  )

  step("Bnest preserves storage configuration and tracked placeholders", context,
    do: outcome(context, :config_and_placeholders_preserved)
  )

  defp prepare(context, state), do: context.behaviour_driver.prepare_behaviour(context, state, [])

  defp perform(context, action),
    do: context.behaviour_driver.perform_behaviour(context, action, [])

  defp outcome(context, expected) do
    assert context.behaviour_driver.behaviour_outcome?(context, expected, [])
    context
  end
end
